### DISCUSSION TITLE
Transparency in Fringes for Blocks

### DISCUSSION THREAD

**Discussion by adimit** [minad/org-modern#178] [Q&A] [answered]
Given the following minimal config:

```elisp
(package-initialize)
(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)
(load-theme 'modus-vivendi)
(add-to-list 'default-frame-alist '(alpha-background . 10))
(use-package org-modern
  :ensure t
  :config
  (global-org-modern-mode))
```

and a recent Emacs compiled for transparency (That'd be [Emacs 29](https://www.emacswiki.org/emacs/TransparentEmacs)—on Wayland, you'll need PGTK enabled), org blocks have a fringe with a solid background:

![screenshot_2024-02-28T22:27+01:00](https://github.com/minad/org-modern/assets/15893/c9c187fe-bbfc-47af-bacf-84df5369bf2e)

Note how the rest of the fringe *does* have a slightly different background, but it's still transparent.

Transparent fringes *are possible*, though I'm not sure if that works with custom bitmaps. Here's an example of an edebug session with a transparent breakpoint indicator:

![screenshot_2024-02-28T22:32+01:00](https://github.com/minad/org-modern/assets/15893/5f99d02a-034a-4715-ba0a-d6af21e7eb72)
 
I tried solving the issue on my own, but I'm stuck. I tried defining a different face for the fringe than `org-block-begin-line`, but even when I left out the `:background` completely, it still showed.

The only workaround I found is to redefine the custom fringe-bitmaps to have width 1. Then only the actual line is shown and the rest of the fringe remains transparent. Of course, we miss out on the pretty corners.

Thanks for `org-modern`, I really love it :)

**Comment by minad** [comment]
Thanks. I don't use Pgtk and Wayland so I cannot easily debug this myself. It could be that this is a Pgtk problem.

Fringe bitmaps are bitmaps, so I think the definitions should not make much of a difference. You can compare the fringe bitmap definitions here:

https://github.com/minad/org-modern/blob/2ab56bdf6dda272a0b7655fb45a2f3f7cd0feb81/org-modern.el#L633-L638

and in Edebug: https://github.com/emacs-mirror/emacs/blob/647cecc853e53a3be0bb2cf5328cd19e677217c9/lisp/emacs-lisp/edebug.el#L3197-L3198

Looking at the face definitions seems to be the correct approach. Maybe try a different face like edebug-enabled-breakpoint? See Edebug: https://github.com/emacs-mirror/emacs/blob/647cecc853e53a3be0bb2cf5328cd19e677217c9/lisp/emacs-lisp/edebug.el#L3228

It could also be that the `wrap-prefix` we use in Org-modern exposes a bug in Emacs itself. As you see, in Edebug, the fringe bitmap is added to an overlay, while we add it to a `wrap-prefix`. Maybe this is the issue? You could check this by using the `edebug-breakpoint` bitmap and `edebug-enabled-breakpoint` face in Org-modern. Is the fringe transparent if you do that?

**Reply by adimit** [reply]
Thanks for your suggestions.

Using a different face didn't work, and even using edebug bitmaps in org-modern shows a black background. So it's more likely to be `wrap-prefix` or something like that. I'll try playing around with setting the fringe in different ways.

I've recompiled Emacs without PGTK and tried it under X, where it is also the case, so the problem isn't Wayland-specific. It may be Linux-specific, as I have no way of checking this on Windows or MacOS.

**Reply by minad** [reply]
Okay, thanks for reporting back! Then this is an issue in Emacs itself and should be reported there via `M-x report-emacs-bug` ideally with a small code snippet demonstrating the issue and the difference of overlays and `wrap-prefix` or `line-prefix`.

**Reply by adimit** [reply]
Okay, I'll try to distill this into something actionable and file it as a bug.

Apropos of nothing, timestamps exhibit the same problems, it seems. 

![screenshot_2024-02-29T17:26+01:00](https://github.com/minad/org-modern/assets/15893/449a4b8c-7e9c-4b90-8304-cc95414cca01)

The black border goes away on hover with the mouse pointer.

**Reply by adimit** [reply]
Should anyone be curious, here's the [bug report](https://lists.gnu.org/archive/html/bug-gnu-emacs/2024-05/msg00015.html) on the emacs-bugs list.

**Reply by adimit** [reply, marked as answer]
The bug [has been fixed](https://lists.gnu.org/archive/html/bug-gnu-emacs/2024-07/msg00386.html) and should be part of the next Emacs release. (Emacs 30)
